### PR TITLE
Fix #8933: Only add the base statistic core if it hasn't already been added

### LINKED ISSUE #8933: Solr statistics core for the current year is queried twice if sharding is enabled, resulting in wrong statistics being displayed
**Describe the bug**
Queries against the statistics core (e.g. for download statistics) are showing the wrong number.
The cause of this seems to be sharding by year: The Solr query contains a list of which cores to include (in this case: all of the sharded ones), but the list includes the current year twice.

The initialization responsible for this is here:

https://github.com/DSpace/DSpace/blob/2b42811e971853c85caf8487411de79515e2bcf3/dspace-api/src/main/java/org/dspace/statistics/SolrLoggerServiceImpl.java#L1686-L1698

The problem is, that `statCoreNames` already includes the core for the current year, because it of course also starts with the base name:

https://github.com/DSpace/DSpace/blob/2b42811e971853c85caf8487411de79515e2bcf3/dspace-api/src/main/java/org/dspace/statistics/SolrLoggerServiceImpl.java#L1681-L1683

**To Reproduce**
1. Enable statistics sharding, set `usage-statistics.shardedByYear = true` in the configuration
2. View usage statistics of a bitstream
3. Download numbers for the current year are counted twice

**Expected behavior**
Download numbers should be accurate

**Related work**
Link to any related tickets or PRs here.

### PR DESCRIPTION
**Pull request by hutattedonmyarm**
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #8933

## Description
Fixes an issue, where statistics for the current year were counted twice, because the solr core was specified twice.

## Instructions for Reviewers
1. Enable statistics sharding, set usage-statistics.shardedByYear = true in the configuration
2. View usage statistics of a bitstream
3. Download a bitstream

Before the fix, this will increase your download numbers by 2. After the fix, it will increase by only 1.

Another good way is to check the solr logs: Sharded queries will have the current year core specified twice as parameter before the PR and only once after.

List of changes in this PR:
* First, ...
* Second, ...

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [ ] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/statistics/SolrLoggerServiceImpl.java [hunk 1]
```diff
@@ -1691,6 +1691,14 @@ protected synchronized void initSolrYearCores() {
                 statisticYearCores
                     .add(baseSolrUrl.replace("http://", "").replace("https://", "") + statCoreName);
             }
+            var baseCore = ((HttpSolrClient) solr)
+                    .getBaseURL()
+                    .replace("http://", "")
+                    .replace("https://", "");
+            if (!statisticYearCores.contains(baseCore)) {
+                //Also add the core containing the current year, if it hasn't been added already
+                statisticYearCores.add(baseCore);
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/statistics/SolrLoggerServiceImpl.java:1700]: @hutattedonmyarm : I'm having a hard time understanding this code change.  It appears (to me) that this `baseCore` will now be added twice.  It's added here on line 1700... but then the same logic appears again on lines 1702-1706 just below this?

I don't understand the benefit of adding it two times... or maybe this code is in the wrong place?

**Review by tdonohue — COMMENTED** [review]
@hutattedonmyarm : I finally got to this PR today, but it's confusing to me.  I'm not exactly sure how to test it unless I already have shards.  The code also seems odd as it looks (to me) like this PR is adding duplicate logic in into the class.

Could you take a look at my inline comment below and help me understand how this is supposed to work?

File: dspace-api/src/main/java/org/dspace/statistics/SolrLoggerServiceImpl.java (hunk 1 above)
Comment (hutattedonmyarm) [reply dspace-api/src/main/java/org/dspace/statistics/SolrLoggerServiceImpl.java:1700]: You're correct, the lines below need to be removed. I just copied the change from out internal patch and overlooked that part. I'll update the PR in a moment.

I *think* you can test it without previous shards, but you need them enabled. If you enable shards, the base core should be added twice (before the PR) and only once (after the PR), even if no shards from previous years exist

**Review by tdonohue — APPROVED** [review]
👍 Thanks @hutattedonmyarm !  This now makes perfect sense after you cleaned it up.

I was able to test this, but had to perform two config changes:
* Had to add `usage-statistics.shardedByYear = true` to my local.cfg
* Had to add `SOLR_OPTS="$SOLR_OPTS -Dsolr.disable.shardsWhitelist=true"` environment variable to tell Solr not to restrict which shards could be queried.  (Otherwise Solr was throwing an error when DSpace tried to query across shards.)

After making those two changes, I could see this work as you describe.  Also verified that with `shardedByYear=false` (the default), everything still works as expected.

**Pull request merged**